// react-use-firebase-hook/dist/react-use-firebase-hook.esm.js
import React, { createContext, useReducer, useEffect, useContext } from 'react';
import { initializeApp, auth } from 'firebase/app';
import 'firebase/auth';

var FirebaseProviderContext =
/*#__PURE__*/
createContext({});

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

var initialState = {
  loading: true,
  firebase: null,
  user: null
};
var reducer = function reducer(state, action) {
  switch (action.type) {
    case 'FIREBASE_LOADED':
      return _extends({}, state, {
        firebase: action.payload
      });

    case 'USER_CHANGE':
      return _extends({}, state, {
        loading: false,
        user: action.payload
      });

    default:
      throw new Error('Invalid action');
  }
};

var FirebaseProvider = function FirebaseProvider(props) {
  var _useReducer = useReducer(reducer, initialState),
      state = _useReducer[0],
      dispatch = _useReducer[1];

  useEffect(function () {
    // Initialize Firebase app on mount and config changes
    var firebase = initializeApp(props.config);
    dispatch({
      type: 'FIREBASE_LOADED',
      payload: firebase
    });
  }, [props.config]);
  useEffect(function () {
    // Check if Firebase instance is valid
    if (!state.firebase) {
      return;
    } // Subscribe to user auth changes


    var unsubscribe = auth().onAuthStateChanged(function (user) {
      dispatch({
        type: 'USER_CHANGE',
        payload: user
      });
    }); // Cleanup subscription on unmount

    return function () {
      return unsubscribe();
    };
  }, [state.firebase]); // If Firebase isn't ready return null

  if (!state.firebase) {
    return null;
  }

  return React.createElement(FirebaseProviderContext.Provider, {
    value: {
      firebase: state.firebase,
      user: state.user,
      loading: state.loading
    }
  }, props.children);
};

var useFirebase = function useFirebase() {
  return useContext(FirebaseProviderContext);
};

export { FirebaseProvider, FirebaseProviderContext, useFirebase };
